### PR TITLE
Log rotation

### PR DESCRIPTION
**Pull request by Legonzaur**
In following to #82, #76 and #73 

Removes .txt files and .dmp files older than x days (7 by default) based on file name

### CODE REVIEW

**Comment by GeckoEidechse** [comment]
Personally I'd go for a higher value like 30 days before deleting logs (the size they accumulate is not very big) to ensure we don't lose any potentially relevant logs (beyond that we probably already release a new version of Northstar anyway, making old logs obsolete).

**Comment by GeckoEidechse** [comment]
Also to avoid any regressions this time, let's test this with
- [ ] log directory missing
- [ ] device current time set to 1st Jan 2022
- [ ] device current time set to 31th December 2022
- [ ] ???
to catch possible edge cases ^^

File: NorthstarDedicatedTest/logging.cpp [hunk 1]
```diff
@@ -10,10 +10,12 @@
 #include <Psapi.h>
 #include <minidumpapiset.h>
 #include "configurables.h"
+#include "logrotation.h"
 
 // This needs to be called after hooks are loaded so we can access the command line args
 void CreateLogFiles()
 {
+	logRotation(7);
```
Comment (pg9182) [inline NorthstarDedicatedTest/logging.cpp:18]: This should not happen if logs are disabled, and there should also be a separate option to disable log rotation (possibly passing it the number of days, with 0 being disabled).

File: NorthstarDedicatedTest/logrotation.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#include "pch.h"
+#include <filesystem>
+#include <string>
+#include <time.h>
+#include <iostream>
+#include <regex>
+#include "configurables.h"
+
+namespace fs = std::filesystem;
+using namespace std;
```
Comment (pg9182) [inline NorthstarDedicatedTest/logrotation.cpp:10]: ^ ...

File: NorthstarDedicatedTest/logrotation.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+double getLogAge(const fs::path path)
+{
+	time_t logtime;
+	time_t currenttime;
+	struct tm timeinfo = {0};
+	int year, month, day;
+	string filename = path.filename().string();
+	// Check if log has a correct name
+	if (regex_search(filename, regex("nslog([0-9]{2}-){2}[0-9]{4}\ [0-9]{2}(-[0-9]{2}){2}.txt")))
```
Comment (pg9182) [inline NorthstarDedicatedTest/logrotation.cpp:22]: The regular expressions should be compiled ahead of time, or at least only once.

File: NorthstarDedicatedTest/logrotation.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		return 0;
+	}
+
+	// Only reads days
+	timeinfo.tm_year = year - 1900;
+	timeinfo.tm_mon = month - 1;
+	timeinfo.tm_mday = day;
+
+	logtime = mktime(&timeinfo);
+	time(&currenttime);
```
Comment (pg9182) [inline NorthstarDedicatedTest/logrotation.cpp:55]: I would personally just return the time, then do the comparison in logRotation.

File: NorthstarDedicatedTest/logrotation.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	return difftime(currenttime, logtime);
+}
+
+void logRotation(int days = 7)
```
Comment (pg9182) [inline NorthstarDedicatedTest/logrotation.cpp:59]: This is incorrect; default arguments belong in the declaration (i.e., the header).

File: NorthstarDedicatedTest/logrotation.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+{
+	string path = GetNorthstarPrefix() + "/logs";
+	struct stat info;
+	// Check if we can access log directory
+	if (stat(path.c_str(), &info) != 0 || !(info.st_mode & S_IFDIR))
+	{
+		return;
+	}
```
Comment (pg9182) [inline NorthstarDedicatedTest/logrotation.cpp:67]: Just check if the directory_iterator is valid rather than doing a separate check.

**Review by pg9182 — CHANGES_REQUESTED** [review]
I don't personally think Northstar should handle log rotation, but if it must, there are a few problems with this implementation.

**Comment by GeckoEidechse** [comment]
> I don't personally think Northstar should handle log rotation, but if it must, there are a few problems with this implementation.

Do it in Viper/VTOL instead?

**Comment by Legonzaur** [comment]
Thank you for your advice, I'll see what I can do on Viper or VTOL instead then

File: NorthstarDedicatedTest/logrotation.cpp (hunk 2 above)
Comment (abarichello) [reply NorthstarDedicatedTest/logrotation.cpp:10]: I think what pg meant is that these namespace shortening are helpful but they shouldn't really be introduced in the code base since it's a Good Practice ™️
